zonemaster/zonemaster-engine
# Add Connectivity04 implementation, and cleanups for Connectivity03

## Purpose

This PR proposes an implementation for the newly specified [Connectivity04 ](https://github.com/zonemaster/zonemaster/blob/develop/docs/public/specifications/tests/Connectivity-TP/connectivity04.md) Test case, as well as various fixes for the Connectivity Test module. It also fixes [missing return codes](https://github.com/zonemaster/zonemaster-engine/pull/1208#issuecomment-1456369848) from the ASNLookup module.

## Context

https://github.com/zonemaster/zonemaster/pull/1149

## Changes

`lib/Zonemaster/Engine/Test/Connectivity.pm` + `profile.json`
- Add implementation for Connectivity04
- Add documentation for Connectivity04
- Fix implementation for Connectivity03 

`lib/Zonemaster/Engine/ASNLookup.pm`
- Update implementation
- Update documentation

`docs/logentry_args.md`
- Update message arguments list

`t/Test-connectivity04.t` + `t/Test-connectivity04.data`
- Add unitary tests for Connectivity04

`t/Test-connectivity03.t` + `t/Test-connectivity03.data`
- Migrate unitary tests from `t/Test-connectivity.t` for Connectivity03

## How to test this PR

Unitary tests should pass. Note that two message tags are not yet tested for this Test Case (see [TODO section](https://github.com/zonemaster/zonemaster-engine/pull/1219/files#diff-da96863768c8affcda3b7a897c4fa641460468d8cb8ec2ef2be5d579126d3e5d) in `t/Test-connectivity4.t`).

Also:
```
$ zonemaster-cli --show-testcase --test connectivity/connectivity04 zonemaster.net --level INFO
Seconds Level     Testcase       Message
======= ========= ============== =======
   0.00 INFO      UNSPECIFIED    Using version v4.6.2 of the Zonemaster engine.
   1.73 INFO      CONNECTIVITY04 The following name server(s) are announced in unique IPv4 prefix(es): "ns2.nic.fr/192.93.0.4;nsa.dnsnode.net/194.58.192.46;nsp.dnsnode.net/194.58.198.32;nsu.dnsnode.net/185.42.137.98"
   1.74 INFO      CONNECTIVITY04 The following name server(s) are announced in unique IPv6 prefix(es): "ns2.nic.fr/2001:660:3005:1::1:2;nsa.dnsnode.net/2a01:3f1:46::53;nsp.dnsnode.net/2a01:3f1:3032::53;nsu.dnsnode.net/2a01:3f0:400::32"

$ zonemaster-cli --show-testcase --test connectivity/connectivity04 zut-root.rd.nic.fr --level INFO
Seconds Level     Testcase       Message
======= ========= ============== =======
   0.00 INFO      UNSPECIFIED    Using version v4.6.2 of the Zonemaster engine.
   1.71 WARNING   CONNECTIVITY04 The following name server(s) are announced in the same IPv4 prefix (145.239.0.0/16): "vps430860.ovh.net/145.239.76.200;vps430862.ovh.net/145.239.76.202"
```
